Update Web Animations keyframe definitions

birtles opened the pull request:
This is an initial attempt at fixing https://github.com/Microsoft/TypeScript/issues/26073

It should possibly be split up into several patches/PRs but I wanted to get some initial feedback first.

A few notes:
* I've included the update to the `CompositeOperationOrAuto` type which was [recent spec change](https://github.com/w3c/csswg-drafts/commit/ced6b5aac02e0e021001f10e4d63e03456686dad), simply because it makes the index type for `PropertyIndexedKeyframes` a little more sane.
* I have no idea how to represent this sort of open dictionary with a few typed keys in TypeScript so this approach might be completely wrong. Bear in mind that with the arrival of animatable custom properties the set of properties that can be specified on a keyframe is essentially open-ended.
* Allowing `number` for the property values is actually ok since WebIDL will perform `ToString` on passed-in numbers so it is actually quite valid (and common) to specify a keyframe as `{ opacity: 0 }` and let WebIDL turn the zero into a string.
* I've made the members of `ComputedKeyframe` required since these dictionary objects are only returned from `getKeyframes()` where they are guaranteed to be filled-in.

Edit: Updated the example about stringification from `{ offset: 0 }` to `{ opacity: 0 }` (since that's what I originally meant to say, particularly since `offset` is actually supposed to be a number).

saschanaz commented:
Using `npm run fetch` is preferred over manually adding types in `addedTypes.json`.

Expected steps:

1. `npm run fetch`
2. Cancel WebGL changes, as it currently causes breakage.
3. `npm run build && npm run baseline-accept`

birtles commented:
Oh, nice! I'll give that a try.

birtles commented:
I've updated and split into two commits: One to run `npm fetch` and update the definitions (except WebGL and Web Speech API since they're broken), and then one to make the changes to Web Animations keyframe definitions.

saschanaz commented on baselines/dom.generated.d.ts:
```diff
@@ -200,11 +200,12 @@ interface ComputedEffectTiming extends EffectTiming {
+    computedOffset: number;
+    easing: string;
+    offset: number | null;
+    [property: string]: string | number | null | undefined;
```
>I have no idea how to represent this sort of open dictionary with a few typed keys

How about using `extends Partial<Record<keyof CSSStyleDeclaration, string | number | null | undefined>>`, while it's somewhat long?

birtles replied:
Does that work for `{ '--variable': '5px' }` though?

saschanaz replied:
Ah, right, custom properties. 😭

saschanaz commented on baselines/dom.generated.d.ts:
```diff
@@ -921,10 +923,11 @@ interface PromiseRejectionEventInit extends EventInit {
+    composite?: CompositeOperationOrAuto | CompositeOperationOrAuto[];
     easing?: string | string[];
     offset?: number | (number | null)[];
+    [property: string]: string | string[] | number | null | (number | null)[] | undefined;
```
Could you explain the `(number | null)[]` thing? Do we also need `(string | null)[]`?

saschanaz replied:
> the type of indexed properties can't conflict with the type of other properties specified on the interface

Ah, that's right. Thanks!

saschanaz commented on inputfiles/overridingTypes.json:
```diff
@@ -2527,15 +2527,41 @@
+                    "member": {
+                        "offset": {
+                            "name": "offset",
+                            "required": 1
```
Makes sense!

You don't need the `name` field anyway, it's only for adding new members.

birtles replied:
Great, thanks. Should be fixed now.

birtles commented:
> LGTM with the name field nits. Probably @mhegazy will review and merge this later, but unfortunately that may take some time as it seems he is not active here these days.

Thanks for the super quick review!

That's fine, it will probably take my employer a few days to approve the CLA anyway.

birtles commented:
I got approval to sign the CLA so this should be all good to go now. I'd recommend NOT squashing and merging since I've split into two commits: (1) to just do the `npm run fetch` and (2) to do the Web Animations-specific changes. Both of them are atomic in that the tests pass after each step.

birtles commented:
@mhegazy review ping?

This is blocking anyone using Web Animations from updating to TS3.

saschanaz commented:
Or maybe @RyanCavanaugh, as I think @mhegazy is on vacation.

birtles commented:
@mhegazy? @RyanCavanaugh? ping?

birtles commented:
Review ping? It's been ~6 weeks or so.

/@andy-ms Do you know anyone who might be able to review/merge this?

jonchardy commented:
@mhegazy @DanielRosenwasser 

Someone please review and merge this. Two months later and it's still blocking all of us using the Web Animations API from updating to TS3.

DanielRosenwasser commented on inputfiles/idl/Web Animations.widl:
```diff
@@ -93,22 +93,22 @@ interface KeyframeEffect : AnimationEffect {
+     double?                  offset = null;
+     double                   computedOffset;
+     DOMString                easing = "linear";
+     CompositeOperationOrAuto composite = "auto";
```
Are these `.widl` files supposed to be hand-edited? Next time anyone runs the `fetcher` script, they'll get overridden, so I would imagine you want to use `overridingTypes.json` to achieve what you want.

birtles replied:
Hi Daniel, thanks for looking at this PR. Please see the comment of this commit (the first of the two). That is, this is literally the result of running the fetcher script so this will not get overridden next time someone runs it.

birtles commented:
Hi @DanielRosenwasser, I responded to your comment but then marked the conversation as resolved so it's possible you may miss it.

I've set this PR up as two commits (which should ideally be merged without squashing):

1. One which updates the definitions using the fetcher script. The exact steps used are described in the extended commit message.
2. One which updates the other files.

DanielRosenwasser commented:
Ah, gotcha! Sorry, it's not totally clear to me and I'm not as familiar with the mechanics of reviewing this repo. Thanks for being patient with me though.

DanielRosenwasser commented:
Just as a heads up, Mohamed is unfortunately no longer our dev lead (:sad:). That's partially why you've seen more delays. Try pinging me in the future though.

birtles commented:
Thanks Daniel! I hope someone can merge this soon.

birtles commented:
@DanielRosenwasser Any chance you could merge this PR? Thanks!

DanielRosenwasser commented:
Thank you for all your patience!
